pkg/sql/schemachanger/scplan/internal/scgraph/dep_edge_tree.go

```diff
@@ -11,100 +11,217 @@
 	"github.com/google/btree"
 )
 
-type depEdgeTree struct {
-	t     *btree.BTree
-	order edgeTreeOrder
-	cmp   nodeCmpFn
-}
-
-type nodeCmpFn func(a, b *screl.Node) (less, eq bool)
-
-func newDepEdgeTree(order edgeTreeOrder, cmp nodeCmpFn) *depEdgeTree {
-	const degree = 8 // arbitrary
-	return &depEdgeTree{
-		t:     btree.New(degree),
-		order: order,
-		cmp:   cmp,
-	}
-}
-
 // edgeTreeOrder order in which the edge tree is sorted,
 // either based on from/to node indexes.
 type edgeTreeOrder bool
 
-func (o edgeTreeOrder) first(e Edge) *screl.Node {
-	if o == fromTo {
-		return e.From()
-	}
-	return e.To()
-}
-
-func (o edgeTreeOrder) second(e Edge) *screl.Node {
-	if o == toFrom {
-		return e.From()
-	}
-	return e.To()
-}
-
 const (
 	fromTo edgeTreeOrder = true
 	toFrom edgeTreeOrder = false
 )
 
-// edgeTreeEntry BTree items for tracking edges
-// in an ordered manner.
-type edgeTreeEntry struct {
-	t    *depEdgeTree
-	edge *DepEdge
+// getTargetIdxFunc is used to get the location of the node in the graph.
+type getTargetIdxFunc = func(n *screl.Node) targetIdx
+
+// depEdges is a data structure to store the set of depEdges. It offers
+// in-order traversal of edges from or to any node in the graph.
+type depEdges struct {
+	fromTo *btree.BTree
+	toFrom *btree.BTree
+
+	getTargetIdx getTargetIdxFunc
+	edgeAlloc    depEdgeAlloc
+	entryAlloc   depEdgeTreeEntryAlloc
+}
+
+// makeDepEdges constructs the depEdge structure.
+func makeDepEdges(getTargetIdx getTargetIdxFunc) depEdges {
+	const degree = 32 // arbitrary
+	return depEdges{
+		fromTo:       btree.New(degree),
+		toFrom:       btree.New(degree),
+		getTargetIdx: getTargetIdx,
+	}
 }
 
-func (et *depEdgeTree) insert(e *DepEdge) {
-	et.t.ReplaceOrInsert(&edgeTreeEntry{
-		t:    et,
-		edge: e,
+// insertOrUpdate will insert a new dep edge if no such edge exists between
+// from and to. Otherwise, it will update the edge accordingly. An error
+// may be returned if the kind is incompatible with the existing kind for
+// the edge.
+func (et *depEdges) insertOrUpdate(rule Rule, kind DepEdgeKind, from, to *screl.Node) error {
+	k := makeEdgeKey(et.getTargetIdx, from, to)
+	if got, ok := et.get(k); ok {
+		return updateExistingDepEdge(rule, kind, got)
+	}
+	de := et.edgeAlloc.new(DepEdge{
+		kind:  kind,
+		from:  from,
+		to:    to,
+		rules: []Rule{rule},
 	})
+	et.fromTo.ReplaceOrInsert(et.entryAlloc.new(k, fromTo, de))
+	et.toFrom.ReplaceOrInsert(et.entryAlloc.new(k, toFrom, de))
+	return nil
 }
 
-func (et *depEdgeTree) get(e *DepEdge) *DepEdge {
-	got, ok := et.t.Get(&edgeTreeEntry{
-		t:    et,
-		edge: e,
-	}).(*edgeTreeEntry)
-	if !ok {
-		return nil
+func updateExistingDepEdge(rule Rule, kind DepEdgeKind, got *DepEdge) error {
+	if got.kind != kind && kind != Precedence {
+		if got.kind != Precedence {
+			return errors.AssertionFailedf(
+				"inconsistent dep edge kinds: %s rule %q conflicts with %s",
+				rule.Kind, rule.Name, got,
+			)
+		}
+		got.kind = kind
 	}
```

Comment:
The second condition check `kind != Precedence` in the `if` is confusing. Why do we need it?

Intuitively, it seems like we just want to update `got.kind` to `kind`, if they are not the same, except that if `got.kind = Precedence` then we throw an error. Thus, I asked the question of why we need `kind != Precedence` here.

Reply:
We also do not want to update it if `kind` is `Precedence` and `got.kind` is not. I added a TODO to handle another case we do not currently handle.